server/utils/match/occupancySummary.ts

```diff
@@ -0,0 +1,76 @@
+import type { Cas1PremiseCapacity, Cas1PremiseCapacityForDay } from '@approved-premises/api'
+import { addDays, differenceInDays } from 'date-fns'
+import { DateFormats, daysToWeeksAndDays } from '../dateUtils'
+
+export type DateRange = {
+  start: string
+  end: string
+}
+
+const daysToRanges = (days: Array<Cas1PremiseCapacityForDay>) =>
+  days.reduce((ranges: Array<DateRange>, capacityForDay) => {
+    if (!ranges.length) {
+      ranges.push({ start: capacityForDay.date, end: capacityForDay.date })
+    } else {
+      const lastRange = ranges[ranges.length - 1]
+      if (differenceInDays(capacityForDay.date, lastRange.end) === 1) {
+        lastRange.end = capacityForDay.date
+      } else {
+        ranges.push({ start: capacityForDay.date, end: capacityForDay.date })
+      }
+    }
+    return ranges
+  }, [])
+
+export const renderDateRange = (dateRange: DateRange): string => {
+  let render = `${DateFormats.isoDateToUIDate(dateRange.start)}`
+  const totalDays = DateFormats.differenceInDays(
+    addDays(DateFormats.isoToDateObj(dateRange.end), 1),
+    DateFormats.isoToDateObj(dateRange.start),
+  ).number
+  const duration = DateFormats.formatDuration(daysToWeeksAndDays(totalDays)).replace(', ', ' and ')
+
+  if (totalDays > 1) {
+    render += ` to ${DateFormats.isoDateToUIDate(dateRange.end)}`
+  }
+
+  render += ` <strong>(${duration})</strong>`
+
+  return render
+}
+
+export const occupancySummary = (premiseCapacity: Cas1PremiseCapacity): string => {
+  const availableDays: Array<Cas1PremiseCapacityForDay> = []
+  const overbookedDays: Array<Cas1PremiseCapacityForDay> = []
+
+  premiseCapacity.capacity.forEach(capacityForDay => {
+    if (capacityForDay.availableBedCount > 0) {
+      availableDays.push(capacityForDay)
+    } else {
+      overbookedDays.push(capacityForDay)
+    }
+  })
+
+  if (availableDays.length === 0) {
+    return `<p class="govuk-heading-m">There are no spaces available for the dates you have selected.</p>`
+  }
+  if (overbookedDays.length === 0) {
+    return `<p class="govuk-heading-m">The placement dates you have selected are available.</p>`
+  }
```

Comment:
this lead to some head scratching but my brain caught up in the end. A coffee and the tests helped the penny drop - can see we are covered :)

Reply:
Yeah, this whole utility might not be hugely optimised (and it certainly won't win any code golf award) but it does the job!

Reply:
it's fine! Just my brain slowly getting into gear after lunch